### PR TITLE
Add AWS lightsail module

### PR DESCRIPTION
**Pull request by NickBall**
##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
lightsail

##### ANSIBLE VERSION

```
ansible 2.2.0.0
  config file = /home/nick/dev/ansible/ansible.cfg
  configured module search path = ['/home/nick/dev/forks/ansible/lib/ansible/modules']
```

##### SUMMARY

New module for provisioning(+removing) VPS instances on the new AWS Lightsail service.

This module follows the same pattern as the ec2 module and reuses the common aws/ec2 argument specs where possible. Also shares a dependency to boto3.

Sample playbook:
```
---
# Provision lightsail instance
- name: Provision lightsail instance
  hosts: localhost
  connection: local
  gather_facts: False
  tasks:
  - name: Launch lightsail instance
    lightsail:
      state: present
      name: myinstance
      region: us-east-1
      zone: us-east-1a
      blueprint_id: ubuntu_16_04
      bundle_id: nano_1_0
      key_pair_name: id_rsa
      wait_timeout: 500
      user_data: " echo 'hello world' > /home/ubuntu/test.txt"
      profile: nick-east-1
    register: my_instance

  - debug:
      msg: "Name is {{ my_instance.instance.name }}"

  - debug:
      var: my_instance

  - name: Delete lightsail instance
    lightsail:
      state: absent
      name: myinstance
      region: us-east-1
      zone: us-east-1a
      profile: nick-east-1
```

Output:
```
PLAY [Provision lightsail instance] ********************************************

TASK [Launch lightsail instance] ***********************************************
changed: [localhost]

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "Name is myinstance"
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "my_instance": {
        "changed": true, 
        "instance": {
            "arn": "arn:aws:lightsail:us-east-1:0000:Instance/0000-1111-2222-3333-4444", 
            "blueprintId": "ubuntu_16_04", 
            "blueprintName": "Ubuntu", 
            "bundleId": "nano_1_0", 
            "createdAt": "2017-01-03T21:34:47.841000-05:00", 
            "hardware": {
                "cpuCount": 1, 
                "ramSizeInGb": 0.5
            }, 
            "isStaticIp": false, 
            "location": {
                "availabilityZone": "us-east-1a", 
                "regionName": "us-east-1"
            }, 
            "name": "myinstance", 
            "networking": {
                "monthlyTransfer": {
                    "gbPerMonthAllocated": 1000
                }, 
                "ports": [
                    {
                        "accessDirection": "inbound", 
                        "accessFrom": "Anywhere (0.0.0.0/0)", 
                        "accessType": "public", 
                        "commonName": "", 
                        "fromPort": 80, 
                        "protocol": "tcp", 
                        "toPort": 80
                    }, 
                    {
                        "accessDirection": "inbound", 
                        "accessFrom": "Anywhere (0.0.0.0/0)", 
                        "accessType": "public", 
                        "commonName": "", 
                        "fromPort": 22, 
                        "protocol": "tcp", 
                        "toPort": 22
                    }
                ]
            }, 
            "privateIpAddress": "172.24.24.24", 
            "resourceType": "Instance", 
            "sshKeyName": "id_rsa", 
            "state": {
                "code": 0, 
                "name": "pending"
            }, 
            "supportCode": "1234/i-5678", 
            "username": "ubuntu"
        }
    }
}

TASK [Delete lightsail instance] ***********************************************
changed: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=4    changed=2    unreachable=0    failed=0   
```

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/lightsail.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/lightsail.py]: Botocore is a dependency of boto3, but this belt-and-suspenders approach of checking both is ok. Can you swap the order so that a user will be prompted to install boto3, which would also resolve a lack of botocore? Basically just swap this conditional with lines 394/395.

File: lib/ansible/modules/cloud/amazon/lightsail.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/lightsail.py:448]: Not sure this part is complete, you may want to be checking for `if 'not found' in e.message: ....` or similar instead of just re-raising.

File: lib/ansible/modules/cloud/amazon/lightsail.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/lightsail.py]: The parens around `changed, instance_dict` are not needed.

File: lib/ansible/modules/cloud/amazon/lightsail.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/lightsail.py]: Please add a message like:

`module.fail_json('Failed while connecting to the lightsail service: %s' % e, exception=traceback.format_exc())`

That way, a prettified traceback can be displayed to the user as well.

Comment (ryansb) [reply lib/ansible/modules/cloud/amazon/lightsail.py]: Also, this will mean you need to `import traceback`

File: lib/ansible/modules/cloud/amazon/lightsail.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/lightsail.py]: Is it possible to narrow the exceptions handled here? This loop could just keep retrying over and over for an error that won't be fixed by waiting longer.

File: lib/ansible/modules/cloud/amazon/lightsail.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/lightsail.py]: For the documentation, you'll also need a `RETURNS` section to show the expected output of this module.

**Review by ryansb — CHANGES_REQUESTED** [review]
This module works, but there are a few new docs & minor changes needed before it can be merged.

File: lib/ansible/modules/cloud/amazon/lightsail.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/lightsail.py]: When exiting, the current module returns keys such as:

```
"instance": {
        "arn": "arn:aws:lightsail:us-east-1:509803855674:Instance/b35a930e-bf27-43bc-9e16-82032254baab", 
        "blueprintId": "ubuntu_16_04", 
        "blueprintName": "Ubuntu", 
        "bundleId": "nano_1_0", 
        "createdAt": "2017-01-20T14:59:15.745000-05:00", 
        "hardware": {
            "cpuCount": 1, 
            "ramSizeInGb": 0.5
        }, 
        "isStaticIp": false, 
        "location": {
            "availabilityZone": "us-east-1a", 
            "regionName": "us-east-1"
        }, 
        "name": "fromansible", 
        "networking": {
.....
```

This return needs to be run through the `camel_dict_to_snake_dict` utility so that values like `isStaticIp` are returned as `is_static_ip` instead.

**Review by ryansb — CHANGES_REQUESTED** [review]

**Comment by s-hertel** [comment]
I tested this module (creating, stopping, starting, rebooting, destroying an instance) successfully using Ansible 2.3.0 on OS-X 10.12.2 with the following playbook:
```
---
- hosts: localhost
  connection: local
  tasks:
  - name: Launch lightsail instance
    lightsail:
      profile: myprofile
      key_pair_name: key-pair
      state: present
      name: myinstance
      region: us-east-1
      zone: us-east-1a
      blueprint_id: ubuntu_16_04
      bundle_id: nano_1_0
      wait_timeout: 200
  - name: stop lightsail instance
    lightsail:
      state: stopped
      name: myinstance
      region: us-east-1
      profile: myprofile
      wait_timeout: 200
  - name: start lightsail instance
    lightsail:
      state: running
      name: myinstance
      region: us-east-1
      profile: myprofile
      wait_timeout: 200
  - name: reboot lightsail instance
    lightsail:
      state: restarted
      name: myinstance
      region: us-east-1
      profile: myprofile
      wait_timeout: 200
  - name: delete lightsail instance
    lightsail:
      state: absent
      name: myinstance
      region: us-east-1
      profile: myprofile
      wait_timeout: 200
```
Result:
```
PLAY [localhost] ******************************************************************************

TASK [Gathering Facts] ************************************************************************
ok: [localhost]

TASK [Launch lightsail instance] **************************************************************
changed: [localhost]

TASK [stop lightsail instance] ****************************************************************
changed: [localhost]

TASK [start lightsail instance] ***************************************************************
changed: [localhost]

TASK [reboot lightsail instance] **************************************************************
changed: [localhost]

TASK [delete lightsail instance] **************************************************************
changed: [localhost]

PLAY RECAP ************************************************************************************
localhost                  : ok=6    changed=5    unreachable=0    failed=0
```

**Comment by NickBall** [comment]
Thanks for the reviews @ryansb , and thanks for testing @s-hertel. I'll circle back and make the requested changes within the next couple of days.

**Comment by ghost** [comment]
so what's the status of this?

**Comment by ryansb** [comment]
Waiting on @NickBall to make changes based on reviews.

**Comment by ghost** [comment]
It appears that @NickBall has been gone for quite a while and shows no signs of returning. Should I open a new issue requesting Lightsail support? This pull request feels like a dead end.

**Comment by s-hertel** [comment]
@ev3y2 Don't open a new issue. We can add more commits by pushing to the cloud_lightsail branch on NickBall/ansible. Sometimes the process for these PRs can be quite slow since community members sometimes cannot dedicate a lot of time. Poking the PR like you've done is a good way to let us know that there is interest. :) I can work on making the requested changes this afternoon and add the commit to this pull request since it looks like it's almost good to go.

**Comment by NickBall** [comment]
Apologies, this slipped off my radar and I haven't had a chance to circle back. I'll get in the changes this weekend, unless @s-hertel gets to it first :-).

**Comment by ghost** [comment]
okay~
thank you @s-hertel and @NickBall

File: lib/ansible/modules/cloud/amazon/lightsail.py
Comment (s-hertel) [reply lib/ansible/modules/cloud/amazon/lightsail.py]: Done.

File: lib/ansible/modules/cloud/amazon/lightsail.py
Comment (s-hertel) [reply lib/ansible/modules/cloud/amazon/lightsail.py:448]: It appears that the exceptions are being handled differently depending on where this function is called, so I added a docstring.

**Comment by ryansb** [comment]
Looks like a few PEP8 issues need fixing before we can get this in.

**Review by ryansb — APPROVED** [review]

**Comment by s-hertel** [comment]
shipit

**Pull request merged**
